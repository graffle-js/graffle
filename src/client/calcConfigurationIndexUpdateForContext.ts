import { hasNonUndefinedKeys } from '../lib/prelude.js'
import type { Configurator } from '../types/configurator.js'
import type { ConfiguratorIndexInput } from '../types/ConfiguratorIndex.js'
import type { Context } from '../types/context.js'

export const calcConfigurationIndexUpdateForContext = <
  context extends Context,
  configurationIndexInput extends ConfiguratorIndexInput,
>(context: context, configurationIndexInput: configurationIndexInput): Context => {
  if (!hasNonUndefinedKeys(configurationIndexInput)) return context
  // todo: performant checking if input changes configuration. If no change, then no copy context.
  // For default input resolvers we can do this automatically (shallow merge)
  // Any custom input resolvers would need to implement their own "is changed" logic.

  const newConfigurationIndex = {
    ...context.configurationIndex,
  }
  for (const configuratorName in configurationIndexInput) {
    const input = configurationIndexInput[configuratorName]!
    const current = context.configurationIndex[configuratorName]!
    const configurator = context.configuratorIndex[configuratorName]!
    const newConfiguration = configurator.inputResolver({ current, input })
    newConfigurationIndex[configuratorName] = newConfiguration
  }
  const newContext = {
    ...context,
    configurationIndex: newConfigurationIndex,
  }
  return newContext
}

export type CalcConfigurationIndexUpdateForContext<
  $Context extends Context,
  $ConfigurationIndexInput extends ConfiguratorIndexInput,
  __ =
    & Omit<$Context, 'configurationIndex'>
    & {
      readonly configurationIndex: {
        readonly [_ in keyof $Context['configurationIndex']]: _ extends keyof $ConfigurationIndexInput
          ? $ConfigurationIndexInput[_] extends object ? Configurator.ApplyInputResolver$Func<
              $Context['configuratorIndex'][_],
              $Context['configurationIndex'][_],
              $ConfigurationIndexInput[_]
            >
          : $Context['configurationIndex'][_]
          : $Context['configurationIndex'][_]
      }
    },
> = __
