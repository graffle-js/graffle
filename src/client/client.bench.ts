import type { ContextEmpty } from '#src/context/ContextEmpty.js'
import { bench } from '@ark/attest'
import { Configuration } from '../context/fragments/configuration/$.js'
import { Transports } from '../context/fragments/transports/$.js'
import type { Client } from './client.js'

bench('Client<ContextEmpty>', () => {
  return {} as Client<ContextEmpty>
}).types([902, 'instantiations'])

bench('Configuration.Add', () => {
  type Result = Client<
    Configuration.Add<ContextEmpty, { output: { defaults: { errorChannel: `return` } } }>
  >
  return {} as Result
}).types([1459, 'instantiations'])

bench('Transports.Configure', () => {
  type Result = Client<
    Transports.Configure<ContextEmpty, 'http', { url: string }>
  >
  return {} as Result
}).types([1007, 'instantiations'])
