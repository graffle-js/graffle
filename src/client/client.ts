import {
  type BuilderExtensionRequestMethods,
  requestMethodsProperties,
} from '../documentBuilder/requestMethods/requestMethods.js' // todo
import { defaultName } from '../generator/config/defaults.js'
import type { Builder } from '../lib/builder/__.js'
import type { ConfigManager } from '../lib/config-manager/__.js'
import { type Exact, proxyGet } from '../lib/prelude.js'
import type { GlobalRegistry } from '../types/GlobalRegistry/GlobalRegistry.js'
import { Schema } from '../types/Schema/__.js'
import { type BuilderExtensionAnyware, builderExtensionAnyware } from './builderExtensions/anyware.js'
import { type BuilderExtensionInternal, builderExtensionInternal } from './builderExtensions/internal.js'
import { type BuilderExtensionScalar, builderExtensionScalar } from './builderExtensions/scalar.js'
import { type BuilderExtensionUse, builderExtensionUse } from './builderExtensions/use.js'
import { type BuilderExtensionWith, builderExtensionWith } from './builderExtensions/with.js'
import { type Context, type ContextWithoutConfig, createContext, type TypeHooksEmpty } from './context.js'
import { type BuilderExtensionGql, builderExtensionGql } from './gql/gql.js'
import { type InputStatic } from './Settings/Input.js'
import { type NormalizeInput } from './Settings/InputToConfig.js'

// export type ClientGeneric = Builder.Definition.MaterializeGeneric<
//   Builder.Definition.ExtendMany<
//     Builder.Definition.Empty,
//     [
//       Internal_,
//       RequestMethods_,
//       With_,
//       Use_,
//       Anyware_,
//       Gql_,
//       Scalar_,
//     ]
//   >
// >

export type Client<$Context extends Context> = Builder.Definition.MaterializeWith<
  ClientDefinition,
  $Context
>

type ClientDefinition = Builder.Definition.Create<[
  BuilderExtensionInternal,
  BuilderExtensionRequestMethods,
  BuilderExtensionWith,
  BuilderExtensionUse,
  BuilderExtensionAnyware,
  BuilderExtensionGql,
  BuilderExtensionScalar,
]>

// dprint-ignore
type Create = <$Input extends InputStatic>(input: Exact<$Input, InputStatic>) =>
  // todo fixme
  // eslint-disable-next-line
  // @ts-ignore
  Client<{
    name: HandleName<$Input>
    input: $Input
    config: NormalizeInput<$Input>
    schemaMap: ConfigManager.OrDefault<$Input['schemaMap'], null>
    retry: null
    extensions: []
    scalars: {}
    typeHooks: TypeHooksEmpty,
  }>

export const create: Create = (input) => {
  const initialContext = createContext({
    name: input.name ?? defaultName,
    schemaMap: input.schemaMap ?? null,
    extensions: [],
    scalars: Schema.Scalar.Registry.empty,
    // retry: null,
    input,
  })
  return createWithContext(initialContext)
}

export const createWithContext = (
  initialContext: ContextWithoutConfig,
) => {
  const context = createContext(initialContext)

  // @ts-expect-error ignoreme
  const clientDirect: Client = {
    ...builderExtensionInternal(createWithContext, context),
    ...builderExtensionGql(createWithContext, context),
    ...builderExtensionWith(createWithContext, context),
    ...builderExtensionUse(createWithContext, context),
    ...builderExtensionAnyware(createWithContext, context),
    ...builderExtensionScalar(createWithContext, context),
  }

  // todo test that access to this works without generation in a unit like test. We discovered bug and covered this in an e2e test.
  Object.assign(clientDirect, {
    ...requestMethodsProperties(createWithContext, context),
  })

  const clientProxy = proxyGet(clientDirect, ({ path, property }) => {
    // eslint-disable-next-line
    // @ts-ignore fixme "Type instantiation is excessively deep and possibly infinite"
    const onGetHandlers = context.extensions.map(_ => _.builder).filter(_ => _ !== undefined)

    for (const onGetHandler of onGetHandlers) {
      const result = onGetHandler.implementation({
        client: clientDirect,
        path,
        property,
      })
      if (result !== undefined) return result
    }

    return undefined
  }) as any

  return clientProxy
}

type HandleName<$Input extends InputStatic> = $Input['name'] extends string ? $Input['name']
  : GlobalRegistry.DefaultClientName
