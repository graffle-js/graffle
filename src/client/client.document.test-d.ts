import { describe, expectTypeOf, test } from 'vitest'
import * as Schema from '../../tests/_/schema/schema.js'
import * as SchemaMutationOnly from '../../tests/_/schemaMutationOnly/schema.js'
import * as SchemaQueryOnly from '../../tests/_/schemaQueryOnly/schema.js'
import { create } from './client.js'

const client = create({ schema: Schema.schema, schemaIndex: Schema.$Index })

test(`requires input`, () => {
  // @ts-expect-error missing input
  client.document()
  // @ts-expect-error empty object
  client.document({})
})

describe(`input`, () => {
  test(`document with one query`, () => {
    const run = client.document({ foo: { query: { id: true } } }).run
    expectTypeOf(run).toMatchTypeOf<(...params: ['foo'] | [] | [undefined]) => Promise<any>>()
  })

  test(`document with two queries`, () => {
    const run = client.document({
      foo: { query: { id: true } },
      bar: { query: { id: true } },
    }).run
    expectTypeOf(run).toMatchTypeOf<(name: 'foo' | 'bar') => Promise<any>>()
  })

  test(`root operation not available if it is not in schema`, () => {
    const clientQueryOnly = create({
      name: `QueryOnly`,
      schema: SchemaQueryOnly.schema,
      schemaIndex: SchemaQueryOnly.$Index,
    })
    clientQueryOnly.document({
      foo: { query: { id: true } },
      // @ts-expect-error mutation not in schema
      bar: { mutation: { id: true } },
    })
    const clientMutationOnly = create({
      name: `MutationOnly`,
      schema: SchemaMutationOnly.schema,
      schemaIndex: SchemaMutationOnly.$Index,
    })
    clientMutationOnly.document({
      // @ts-expect-error query not in schema
      foo: { query: { id: true } },
      bar: { mutation: { id: true } },
    })
  })
})

describe(`output`, () => {
  test(`document with one query`, async () => {
    {
      const result = await client.document({ foo: { query: { id: true } } }).run()
      expectTypeOf(result).toEqualTypeOf<{ id: string | null }>()
    }
    {
      const result = await client.document({ foo: { query: { id: true } } }).run(`foo`)
      expectTypeOf(result).toEqualTypeOf<{ id: string | null }>()
    }
    {
      const result = await client.document({ foo: { query: { id: true } } }).run(undefined)
      expectTypeOf(result).toEqualTypeOf<{ id: string | null }>()
    }
  })
  test(`document with two queries`, async () => {
    const result = await client.document({
      foo: { query: { id: true } },
      bar: { query: { id: true } },
    }).run(`foo`)
    expectTypeOf(result).toEqualTypeOf<{ id: string | null }>()
  })
})
