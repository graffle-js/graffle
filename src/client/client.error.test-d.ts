/* eslint-disable */
import { expectTypeOf, test } from 'vitest'
import { isError } from '../../tests/_/schema/generated/Error.js'
import * as Schema from '../../tests/_/schema/schema.js'
import { create } from './client.js'

const client = create({ schema: Schema.schema, schemaIndex: Schema.$Index })

test('isError utility function narrows for error objects', async () => {
  const result = await client.query.result({ $: { case: 'Object1' }, __typename: true })

  if (isError(result)) {
    expectTypeOf(result).toEqualTypeOf<{ __typename: 'ErrorOne' } | { __typename: 'ErrorTwo' }>()
  } else {
    expectTypeOf(result).toEqualTypeOf<null | { __typename: 'Object1' }>()
  }
})
