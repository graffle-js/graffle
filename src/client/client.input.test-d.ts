import { test } from 'vitest'
import { $Index } from '../../tests/_/schema/generated/SchemaRuntime.js'
import { schema } from '../../tests/_/schema/schema.js'
import { create } from './client.js'

test(`works`, () => {
  create({ schemaIndex: $Index, schema, name: `QueryOnly`, returnMode: `graphql` })
  create({ schemaIndex: $Index, schema, name: `QueryOnly`, returnMode: `data` })
  create({ schemaIndex: $Index, schema, name: `QueryOnly`, returnMode: `dataAndErrors` })
  // @ts-expect-error bad returnMode
  create({ schemaIndex: $Index, schema, name: `QueryOnly`, returnMode: `successData` })

  create({ schemaIndex: $Index, schema, name: `default`, returnMode: `graphql` })
  create({ schemaIndex: $Index, schema, name: `default`, returnMode: `data` })
  create({ schemaIndex: $Index, schema, name: `default`, returnMode: `dataAndErrors` })
  create({ schemaIndex: $Index, schema, name: `default`, returnMode: `successData` })

  create({ schemaIndex: $Index, schema, returnMode: `graphql` })
  create({ schemaIndex: $Index, schema, returnMode: `data` })
  create({ schemaIndex: $Index, schema, returnMode: `dataAndErrors` })
  create({ schemaIndex: $Index, schema, returnMode: `successData` })
})
