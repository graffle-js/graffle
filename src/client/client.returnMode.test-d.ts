/* eslint-disable */
import { ExecutionResult } from 'graphql'
import { ObjMap } from 'graphql/jsutils/ObjMap.js'
import { describe } from 'node:test'
import { expectTypeOf, test } from 'vitest'
import { $Index as schemaIndex } from '../../tests/_/schema/generated/SchemaRuntime.js'
import { schema } from '../../tests/_/schema/schema.js'
import { create } from './client.js'

// dprint-ignore
describe('default', () => {
  const client = create({ schema, schemaIndex })
  test(`document`, async () => {
    expectTypeOf(client.document({ main: { query: { id: true } } }).run()).resolves.toEqualTypeOf<{ id: string | null }>()
  })
  test(`raw`, async () => {
    expectTypeOf(client.raw('query main {\nid\n}', {}, 'main')).resolves.toEqualTypeOf<ExecutionResult>()
  })
  test('query field method', async () => {
    await expectTypeOf(client.query.__typename()).resolves.toEqualTypeOf<'Query'>()
  })
  test('query $batch', async () => {
    await expectTypeOf(client.query.$batch({ __typename: true, id: true })).resolves.toEqualTypeOf<{ __typename: 'Query', id: string|null }>()
  })
})

// dprint-ignore
describe('data', () => {
  const client = create({ schema, schemaIndex, returnMode: 'data' })
  test(`document`, async () => {
    expectTypeOf(client.document({ main: { query: { id: true } } }).run()).resolves.toEqualTypeOf<{ id: string | null }>()
  })
  test(`raw`, async () => {
    expectTypeOf(client.raw('query main {\nid\n}', {}, 'main')).resolves.toEqualTypeOf<ExecutionResult>()
  })
  test('query field method', async () => {
    await expectTypeOf(client.query.__typename()).resolves.toEqualTypeOf<'Query'>()
  })
  test('query $batch', async () => {
    await expectTypeOf(client.query.$batch({ __typename: true, id: true })).resolves.toEqualTypeOf<{ __typename: 'Query', id: string|null }>()
  })
})

// dprint-ignore
describe('graphql', () => {
  const client = create({ schema, schemaIndex, returnMode: 'graphql' })
  test(`document`, async () => {
    expectTypeOf(client.document({ main: { query: { id: true } } }).run()).resolves.toEqualTypeOf<ExecutionResult<{ id: string | null }, ObjMap<unknown>>>()
  })
  test(`raw`, async () => {
    expectTypeOf(client.raw('query main {\nid\n}', {}, 'main')).resolves.toEqualTypeOf<ExecutionResult>()
  })
  test('query field method', async () => {
    await expectTypeOf(client.query.__typename()).resolves.toEqualTypeOf<ExecutionResult<{ __typename: 'Query' }>>()
  })
  test('query $batch', async () => {
    await expectTypeOf(client.query.$batch({ __typename: true, id: true })).resolves.toEqualTypeOf<ExecutionResult<{ __typename: 'Query', id: string|null }>>()
  })
})
