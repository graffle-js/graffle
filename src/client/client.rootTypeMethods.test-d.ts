/* eslint-disable */
import { expectTypeOf, test } from 'vitest'
import * as Schema from '../../tests/_/schema/schema.js'
import { create } from './client.js'

const client = create<Schema.Index>({ schema: Schema.schema, schemaIndex: Schema.$Index })

// dprint-ignore
test(`query`, () => {
  // scalar
  expectTypeOf(client.query.id).toEqualTypeOf<() => Promise<string | null>>()
  expectTypeOf(client.query.idNonNull).toEqualTypeOf<() => Promise<string>>()
  // custom scalar
  expectTypeOf(client.query.date).toEqualTypeOf<() => Promise<Date | null>>()
  expectTypeOf(client.query.dateNonNull).toEqualTypeOf<() => Promise<Date>>()
  expectTypeOf(client.query.dateArg).toMatchTypeOf<(args?: { date?: Date | null }) => Promise<Date | null>>()
  expectTypeOf(client.query.dateArgNonNull).toMatchTypeOf<(args: { date: Date }) => Promise<Date | null>>()
  const x2 = client.query.dateObject1({ date1: true })
  // object
  expectTypeOf(client.query.dateObject1({ date1: true })).resolves.toEqualTypeOf<{ date1: Date | null } | null>()
  expectTypeOf(client.query.dateObject1({ $scalars: true })).resolves.toEqualTypeOf<{ __typename: "DateObject1"; date1: Date | null } | null>()
  expectTypeOf(client.query.unionFooBar({ onFoo: { id: true }})).resolves.toEqualTypeOf<{} | { id: string | null } | null>()
  expectTypeOf(client.query.interface({ id: true })).resolves.toEqualTypeOf<null | { id: string | null }>()
  expectTypeOf(client.query.interface({ onObject1ImplementingInterface: { int: true }})).resolves.toEqualTypeOf<{} | { int: number | null } | null>()

  // @ts-expect-error missing input selection set
  client.query.dateObject1()  
  // @ts-expect-error excess properties
  const x = client.query.dateObject1({ abc: true })  
  // @ts-expect-error no directives on root type object fields
  client.query.dateObject1({ $defer: true })  
  // todo @ts-expect-error empty object
  // client.query.dateObject1({})  
})
