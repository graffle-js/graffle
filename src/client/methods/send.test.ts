/* oxlint-disable no-constant-condition */
import { Graffle } from '#graffle'
import { Var } from '#graffle/client'
import { db } from '#test/schema/possible/db.js'
import { possibleSchema } from '#test/schema/possible/schema.js'
import { Ts } from '@wollybeard/kit'
import { Test } from '@wollybeard/kit/test'
import { describe, expect, test } from 'vitest'
import { Possible } from '../../extensions/DocumentBuilder/__tests__/fixtures/possible/$.js'
import { TransportMemory } from '../../extensions/TransportMemory/TransportMemory.js'
const $ = Var.$
const g = Graffle
  .create()
  .use(TransportMemory)
  .transport(`memory`, { schema: possibleSchema })

Test
  .describe('send() with documents')
  .inputType<{ doc: string; sendArgs: any[] }>()
  .describeInputs('GraphQL Documents', [
    { doc: `{ id }`, sendArgs: [] },
    { doc: `query getID { id } query getIDNonNull { idNonNull }`, sendArgs: ['getID'] },
    { doc: `query getID { id } query getIDNonNull { idNonNull }`, sendArgs: ['getIDNonNull'] },
    { doc: `query getID { id } mutation mutateID { id }`, sendArgs: ['getID'] },
    { doc: `query getID { id } mutation mutateID { id }`, sendArgs: ['mutateID'] },
  ])
  .test(async ({ input: { doc, sendArgs } }) => {
    return await g.send(doc as any, ...sendArgs)
  })

describe('TypedFullDocumentString', () => {
  test('with single operation - no operation name parameter', () => {
    const doc = Possible.document({ query: { foo: { id: true } } })
    // Type checking only
    if (false) {
      // @ts-expect-error - Unreachable code for type testing
      const result = g.send(doc)
      Ts.Test.exact<Promise<{ id: string | null } | null>>()(result)
    }
  })

  test('single operation, optional var -> no name, yes variables optional', async () => {
    const doc = Possible.document({ query: { foo: { stringWithRequiredArg: { $: { string: $.default('abc') } } } } })
    await g.send(doc)
    await g.send(doc, { string: '' })

    // Type checking only - not executed
    if (false) {
      // @ts-expect-error - invalid type
      g.send(doc, { string: 1 })

      Ts.Test.exact<Promise<{ stringWithRequiredArg: string | null } | null>>()(g.send(doc))
    }
  })

  test('single operation, required var -> variables parameter is required', async () => {
    const doc = Possible.document({ query: { foo: { stringWithRequiredArg: { $: { string: $ } } } } })

    // Type checking only - not executed
    if (false) {
      // @ts-expect-error - missing required variables
      g.send(doc)
      // @ts-expect-error - invalid type
      g.send(doc, { string: 1 })
    }

    const result = await g.send(doc, { string: '' })
    expect(result).toEqual({ stringWithRequiredArg: '' })
    Ts.Test.exact<Promise<{ stringWithRequiredArg: string | null } | null>>()(g.send(doc, { string: '' }))
  })

  test('multiple operations, no var -> operation name required', async () => {
    const doc = Possible.document({
      query: {
        a: { id: true },
        b: { idNonNull: true },
      },
    })

    // Type checking only - not executed
    if (false) {
      // @ts-expect-error - missing operation name
      g.send(doc)
      // @ts-expect-error - invalid operation name
      g.send(doc, 'c')
    }

    const resultA = await g.send(doc, 'a')
    const resultB = await g.send(doc, 'b')
    expect(resultA).toEqual({ id: db.id1 })
    expect(resultB).toEqual({ idNonNull: db.id1 })

    Ts.Test.exact<Promise<{ id: string | null } | null>>()(g.send(doc, 'a'))
    Ts.Test.exact<Promise<{ idNonNull: string } | null>>()(g.send(doc, 'b'))
  })

  test('multiple operations, 1 optional var/1 required var -> name required, var optional/var required', async () => {
    const doc = Possible.document({
      query: {
        a: { stringWithRequiredArg: { $: { string: $ } } },
        b: { stringWithRequiredArg: { $: { string: $.default('') } } },
      },
    })

    // Type checking only - not executed
    if (false) {
      // @ts-expect-error - missing operation name
      g.send(doc)
      // @ts-expect-error - invalid operation name
      g.send(doc, 'c')
      // @ts-expect-error - missing variable arguments
      g.send(doc, 'a')
    }

    const resultA = await g.send(doc, 'a', { string: '' })
    expect(resultA).toEqual({ stringWithRequiredArg: '' })

    // Type checking only - not executed
    if (false) {
      // @ts-expect-error - Unreachable code for type testing
      const resultB = g.send(doc, 'b', { string: '' })
      const resultBDefault = g.send(doc, 'b')

      Ts.Test.exact<Promise<{ stringWithRequiredArg: string | null } | null>>()(g.send(doc, 'a', { string: '' }))
      Ts.Test.exact<Promise<{ stringWithRequiredArg: string | null } | null>>()(resultB)
      Ts.Test.exact<typeof resultB>()(resultBDefault)
    }
  })

  test('send() with mixed query/mutation document and variable inference', () => {
    const doc = Possible.document({
      query: {
        a: {
          objectWithArgs: {
            $: {
              id: $.required(),
              string: $.required(),
            },
            id: true,
          },
        },
      },
      mutation: {
        b: {
          // Mutation fields in Possible schema don't accept arguments
          idNonNull: true,
        },
      },
    })

    // Type checking only - not executed
    if (false) {
      // @ts-expect-error - Unreachable code for type testing
      const client = Possible.create({ check: { preflight: false } })

      // Should accept 'id' and 'string' variables
      client.send(doc, 'a', { id: 'foo', string: 'bar' })

      // Mutation operation with no variables works
      client.send(doc, 'b')
    }
  })

  // TODO: Plain string support
  // test('plain string - no type safety', () => {
  //   const query = `{ foo { id } }`
  //   const result = client.send(query)
  //   Ts.Test.exact<Promise<{ [x: string]: any } | null>>()(result)
  // })
})
