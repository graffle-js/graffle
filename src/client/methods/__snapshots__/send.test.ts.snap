// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`GraphQL Documents > send() with documents > fn({ doc: '{ id }', sendArgs: [] }) 1`] = `
"
╔══════════════════════════════════════════════════╗ GIVEN ARGUMENTS
{
  doc: '{ id }',
  sendArgs: []
}
╠══════════════════════════════════════════════════╣ THEN RETURNS OBJECT
[Object: null prototype] {
  id: 'abc'
}
╚══════════════════════════════════════════════════╝"
`;

exports[`GraphQL Documents > send() with documents > fn({ doc: 'query getID { id } mutation mutateID { id }', sendArgs: [ 'getID' ] }) 1`] = `
"
╔══════════════════════════════════════════════════╗ GIVEN ARGUMENTS
{
  doc: 'query getID { id } mutation mutateID { id }',
  sendArgs: [ 'getID' ]
}
╠══════════════════════════════════════════════════╣ THEN RETURNS OBJECT
[Object: null prototype] {
  id: 'abc'
}
╚══════════════════════════════════════════════════╝"
`;

exports[`GraphQL Documents > send() with documents > fn({ doc: 'query getID { id } mutation mutateID { id }', sendArgs: [ 'mutateID' ] }) 1`] = `
"
╔══════════════════════════════════════════════════╗ GIVEN ARGUMENTS
{
  doc: 'query getID { id } mutation mutateID { id }',
  sendArgs: [ 'mutateID' ]
}
╠══════════════════════════════════════════════════╣ THEN RETURNS OBJECT
[Object: null prototype] {
  id: 'abc'
}
╚══════════════════════════════════════════════════╝"
`;

exports[`GraphQL Documents > send() with documents > fn({ doc: 'query getID { id } query getIDNonNull { idNonNull }', sendArgs: [ 'getID...) 1`] = `
"
╔══════════════════════════════════════════════════╗ GIVEN ARGUMENTS
{
  doc: 'query getID { id } query getIDNonNull { idNonNull }',
  sendArgs: [ 'getID' ]
}
╠══════════════════════════════════════════════════╣ THEN RETURNS OBJECT
[Object: null prototype] {
  id: 'abc'
}
╚══════════════════════════════════════════════════╝"
`;

exports[`GraphQL Documents > send() with documents > fn({ doc: 'query getID { id } query getIDNonNull { idNonNull }', sendArgs: [ 'getID...) 2`] = `
"
╔══════════════════════════════════════════════════╗ GIVEN ARGUMENTS
{
  doc: 'query getID { id } query getIDNonNull { idNonNull }',
  sendArgs: [ 'getIDNonNull' ]
}
╠══════════════════════════════════════════════════╣ THEN RETURNS OBJECT
[Object: null prototype] {
  idNonNull: 'abc'
}
╚══════════════════════════════════════════════════╝"
`;
